src/algorithms/matching.js

```diff
@@ -0,0 +1,1257 @@
+        // Decide in which direction we will go round the blossom.
+        j = blossomchilds[b].indexOf(entrychild);
+        if (j & 1) {
+          // Start index is odd; go forward.
+          jstep = 1;
+          endptrick = 0;
+          stop = blossomchilds[b].length;
+          base = 0;
+        }
+        else {
+          // Start index is even; go backward.
+          jstep = -1;
+          endptrick = 1;
+          stop = 0;
+          base = blossomchilds[b].length;
+        }
+        // Move along the blossom until we get to the base.
+        p = labelend[b];
+        while (j !== stop) {
+          // Relabel the T-sub-blossom.
+          label[endpoint[p ^ 1]] = 0;
+          label[endpoint[blossomendps[b][j-endptrick]^endptrick^1]] = 0;
+          assignLabel(endpoint[p ^ 1], 2, p);
+          // Step to the next S-sub-blossom and note its forward endpoint.
+          allowedge[Math.floor(blossomendps[b][j-endptrick]/2)] = true;
+          j += jstep;
+          p = blossomendps[b][j-endptrick] ^ endptrick;
+          // Step to the next T-sub-blossom.
+          allowedge[Math.floor(p/2)] = true;
+          j += jstep;
+        }
+        // Relabel the base T-sub-blossom WITHOUT stepping through to
+        // its mate (so don't call assignLabel).
+        bv = blossomchilds[b][0];
+        label[endpoint[p ^ 1]] = label[bv] = 2;
+        labelend[endpoint[p ^ 1]] = labelend[bv] = p;
+        bestedge[bv] = -1;
+        // Continue along the blossom until we get back to entrychild.
+        j = base + jstep;
+        while (blossomchilds[b][j] !== entrychild) {
+          // Examine the vertices of the sub-blossom to see whether
+          // it is reachable from a neighbouring S-vertex outside the
+          // expanding blossom.
+          bv = blossomchilds[b][j];
+          if (label[bv] === 1) {
+            // This sub-blossom just got label S through one of its
+            // neighbours; leave it.
+            j += jstep;
+            continue;
+          }
+          blossomLeaves(bv, function(v){
+            if (label[v] !== 0) {
+              // If the sub-blossom contains a reachable vertex, assign
+              // label T to the sub-blossom.
+              assert(label[v] === 2);
+              assert(inblossom[v] === bv);
+              label[v] = 0;
+              label[endpoint[mate[blossombase[bv]]]] = 0;
+              assignLabel(v, 2, labelend[v]);
+              return true;
+            }
+          });
+
+          j += jstep;
+        }
+      }
+      // Recycle the blossom number.
+      label[b] = labelend[b] = -1;
+      blossomchilds[b] = blossomendps[b] = null;
+      blossombase[b] = -1;
+      blossombestedges[b] = null;
+      bestedge[b] = -1;
+      unusedblossoms.push(b);
+    };
+
+    var rotate = function (a, n) {
+      var head = a.splice(0, n);
+      for (var i = 0; i < n; ++i) {
+        a.push(head[i]);
+      }
+    };
+
```

Comment:
It would be more efficient to use deques here, is there an available implementation?